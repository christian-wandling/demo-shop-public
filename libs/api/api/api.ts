export * from './cart-items-api';
import { CartItemsApi } from './cart-items-api';
export * from './orders-api';
import { OrdersApi } from './orders-api';
export * from './products-api';
import { ProductsApi } from './products-api';
export * from './shopping-sessions-api';
import { ShoppingSessionsApi } from './shopping-sessions-api';
export * from './users-api';
import { UsersApi } from './users-api';
export const APIS = [CartItemsApi, OrdersApi, ProductsApi, ShoppingSessionsApi, UsersApi];
