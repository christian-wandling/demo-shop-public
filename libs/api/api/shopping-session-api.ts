/**
 * Demo Shop API
 *
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/* tslint:disable:no-unused-variable member-ordering */

import { Inject, Injectable, Optional } from '@angular/core';
import {
  HttpClient,
  HttpHeaders,
  HttpParams,
  HttpResponse,
  HttpEvent,
  HttpParameterCodec,
  HttpContext,
} from '@angular/common/http';
import { CustomHttpParameterCodec } from '../encoder';
import { Observable } from 'rxjs';

// @ts-ignore
import { AddCartItem401Response } from '../model/add-cart-item401-response';
// @ts-ignore
import { AddCartItemRequest } from '../model/add-cart-item-request';
// @ts-ignore
import { CartItemResponse } from '../model/cart-item-response';
// @ts-ignore
import { OrderResponse } from '../model/order-response';
// @ts-ignore
import { ShoppingSessionResponse } from '../model/shopping-session-response';
// @ts-ignore
import { UpdateCartItemQuantityRequest } from '../model/update-cart-item-quantity-request';
// @ts-ignore
import { UpdateCartItemQuantityResponse } from '../model/update-cart-item-quantity-response';
// @ts-ignore
import { ValidationProblemDetails } from '../model/validation-problem-details';

// @ts-ignore
import { BASE_PATH, COLLECTION_FORMATS } from '../variables';
import { Configuration } from '../configuration';

@Injectable({
  providedIn: 'root',
})
export class ShoppingSessionApi {
  protected basePath = 'http://localhost';
  public defaultHeaders = new HttpHeaders();
  public configuration = new Configuration();
  public encoder: HttpParameterCodec;

  constructor(
    protected httpClient: HttpClient,
    @Optional() @Inject(BASE_PATH) basePath: string | string[],
    @Optional() configuration: Configuration
  ) {
    if (configuration) {
      this.configuration = configuration;
    }
    if (typeof this.configuration.basePath !== 'string') {
      const firstBasePath = Array.isArray(basePath) ? basePath[0] : undefined;
      if (firstBasePath != undefined) {
        basePath = firstBasePath;
      }

      if (typeof basePath !== 'string') {
        basePath = this.basePath;
      }
      this.configuration.basePath = basePath;
    }
    this.encoder = this.configuration.encoder || new CustomHttpParameterCodec();
  }

  // @ts-ignore
  private addToHttpParams(httpParams: HttpParams, value: any, key?: string): HttpParams {
    if (typeof value === 'object' && value instanceof Date === false) {
      httpParams = this.addToHttpParamsRecursive(httpParams, value);
    } else {
      httpParams = this.addToHttpParamsRecursive(httpParams, value, key);
    }
    return httpParams;
  }

  private addToHttpParamsRecursive(httpParams: HttpParams, value?: any, key?: string): HttpParams {
    if (value == null) {
      return httpParams;
    }

    if (typeof value === 'object') {
      if (Array.isArray(value)) {
        (value as any[]).forEach(elem => (httpParams = this.addToHttpParamsRecursive(httpParams, elem, key)));
      } else if (value instanceof Date) {
        if (key != null) {
          httpParams = httpParams.append(key, (value as Date).toISOString().substring(0, 10));
        } else {
          throw Error('key may not be null if value is Date');
        }
      } else {
        Object.keys(value).forEach(
          k => (httpParams = this.addToHttpParamsRecursive(httpParams, value[k], key != null ? `${key}.${k}` : k))
        );
      }
    } else if (key != null) {
      httpParams = httpParams.append(key, value);
    } else {
      throw Error('key may not be null if value is not object or array');
    }
    return httpParams;
  }

  /**
   * Add cart item
   * Add a cart item to the shopping session of current user based on identity extracted from bearer token
   * @param addCartItemRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<CartItemResponse>;
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<CartItemResponse>>;
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<CartItemResponse>>;
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json', 'text/json', 'application/*+json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current/cart-items`;
    return this.httpClient.request<CartItemResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      body: addCartItemRequest,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Checkout current shopping session
   * Check out by creating an order from the current shopping session and deleting the shopping session afterwardds
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public checkout(
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<OrderResponse>;
  public checkout(
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<OrderResponse>>;
  public checkout(
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<OrderResponse>>;
  public checkout(
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/checkout`;
    return this.httpClient.request<OrderResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Remove cart item
   * Remove a cart item from the shopping session of current user based on identity extracted from bearer token
   * @param id
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public removeCartItem(
    id: number,
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any>;
  public removeCartItem(
    id: number,
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<any>>;
  public removeCartItem(
    id: number,
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<any>>;
  public removeCartItem(
    id: number,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    if (id === null || id === undefined) {
      throw new Error('Required parameter id was null or undefined when calling removeCartItem.');
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current/cart-items/${this.configuration.encodeParam({
      name: 'id',
      value: id,
      in: 'path',
      style: 'simple',
      explode: false,
      dataType: 'number',
      dataFormat: 'int32',
    })}`;
    return this.httpClient.request<any>('delete', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Resolve current shopping session
   * Resolve current shopping session based on identity extracted from bearer token
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public resolveCurrentShoppingSession(
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<ShoppingSessionResponse>;
  public resolveCurrentShoppingSession(
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<ShoppingSessionResponse>>;
  public resolveCurrentShoppingSession(
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<ShoppingSessionResponse>>;
  public resolveCurrentShoppingSession(
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current`;
    return this.httpClient.request<ShoppingSessionResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Update cart item quantity
   * Update quantity of a cart item in shopping session of current user based on identity extracted from bearer token
   * @param id
   * @param updateCartItemQuantityRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<UpdateCartItemQuantityResponse>;
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<UpdateCartItemQuantityResponse>>;
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<UpdateCartItemQuantityResponse>>;
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    if (id === null || id === undefined) {
      throw new Error('Required parameter id was null or undefined when calling updateCartItemQuantity.');
    }
    if (updateCartItemQuantityRequest === null || updateCartItemQuantityRequest === undefined) {
      throw new Error(
        'Required parameter updateCartItemQuantityRequest was null or undefined when calling updateCartItemQuantity.'
      );
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json', 'text/json', 'application/*+json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current/cart-items/${this.configuration.encodeParam({
      name: 'id',
      value: id,
      in: 'path',
      style: 'simple',
      explode: false,
      dataType: 'number',
      dataFormat: 'int32',
    })}`;
    return this.httpClient.request<UpdateCartItemQuantityResponse>(
      'patch',
      `${this.configuration.basePath}${localVarPath}`,
      {
        context: localVarHttpContext,
        body: updateCartItemQuantityRequest,
        responseType: <any>responseType_,
        withCredentials: this.configuration.withCredentials,
        headers: localVarHeaders,
        observe: observe,
        transferCache: localVarTransferCache,
        reportProgress: reportProgress,
      }
    );
  }
}
