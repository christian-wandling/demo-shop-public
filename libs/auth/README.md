# `@demo-shop/auth`

Authentication and authorization library built on Keycloak for Angular applications. This library provides a comprehensive solution for securing your application with role-based access control and seamless authentication flows.
Features

## Features

- **Keycloak Integration**

  - SSO (Single Sign-On) support
  - OAuth2/OpenID Connect compliant
  - Session timeout handling
  - Token refresh handling

- **Permission-based Route Guards**

  - Role-based route protection

- **Authentication Interceptor for HTTP Requests**

  - Authentication status monitoring
  - Automatic token attachment

- **Auth Facade**
  - Abstracting permission and authentication logic

## API Reference

For a complete API reference, see the [Compodoc documentation](https://christian-wandling.github.io/demo-shop-public/compodoc/frontend).

## Running unit tests

Run `nx test auth` to execute the unit tests.
