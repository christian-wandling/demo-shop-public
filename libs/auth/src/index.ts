export * from './lib/auth.facade';
export * from './lib/providers/provide-auth';
export * from './lib/models/auth-config';
export * from './lib/interceptors/auth.interceptor';
export * from './lib/guards/permission.guard';
export * from './lib/services/permission.service';
export * from './lib/enums/permission-strategy';
