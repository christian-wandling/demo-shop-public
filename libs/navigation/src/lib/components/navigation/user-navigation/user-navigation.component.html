<ng-container *ngIf="isSmallScreen(); else largeScreen">
  @if (currentUser()) {
  <div class="flow-root">
    <a class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer">
      {{ currentUser()?.firstname }}&nbsp;{{ currentUser()?.lastname }}
    </a>
  </div>
  <div class="flow-root">
    <a (click)="logout()" class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer"> Sign out </a>
  </div>
  } @else {
  <div class="flow-root">
    <a (click)="login()" class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer"> Sign in </a>
  </div>
  <div class="flow-root">
    <a (click)="register()" class="-m-2 block p-2 font-medium text-gray-900 cursor-pointer"> Register </a>
  </div>
  }
</ng-container>

<ng-template #largeScreen>
  @if (currentUser()) {
  <a class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer">
    {{ currentUser()?.firstname }}&nbsp;{{ currentUser()?.lastname }}
  </a>
  <span class="h-6 w-px bg-gray-200" aria-hidden="true"></span>
  <a (click)="logout()" class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"> Sign out </a>
  } @else {
  <a (click)="login()" class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"> Sign in </a>
  <span class="h-6 w-px bg-gray-200" aria-hidden="true"></span>
  <a (click)="register()" class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"> Register </a>
  }
</ng-template>
