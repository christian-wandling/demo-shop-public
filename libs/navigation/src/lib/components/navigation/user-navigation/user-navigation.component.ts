import { booleanAttribute, ChangeDetectionStrategy, Component, inject, input } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AuthFacade } from '@demo-shop/auth';
import { UserFacade } from '@demo-shop/user';
import { CartFacade } from '@demo-shop/cart';

/**
 * Component that handles user navigation functionality including login, registration, and logout
 *
 * @example
 * <lib-user-navigation isSmallScreen>
 */
@Component({
  selector: 'lib-user-navigation',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './user-navigation.component.html',
  styleUrl: './user-navigation.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class UserNavigationComponent {
  /** Flag indicating if the current view is on a small screen */
  readonly isSmallScreen = input(false, { transform: booleanAttribute });

  readonly #authFacade = inject(AuthFacade);
  readonly #userFacade = inject(UserFacade);
  readonly currentUser = this.#userFacade.getCurrentUser();
  readonly #cartFacade = inject(CartFacade);

  /** Handles user login and loads shopping session */
  async login(): Promise<void> {
    await this.#authFacade.login();
    await this.#cartFacade.loadShoppingSession();
  }

  /** Handles user registration and loads shopping session */
  async register(): Promise<void> {
    await this.#authFacade.register();
    await this.#cartFacade.loadShoppingSession();
  }

  /** Handles user logout */
  async logout(): Promise<void> {
    await this.#authFacade.logout();
  }
}
