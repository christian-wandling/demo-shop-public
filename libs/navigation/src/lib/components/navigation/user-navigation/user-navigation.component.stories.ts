import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule } from '@angular/common';
import { importProvidersFrom, signal } from '@angular/core';
import { UserResponse } from '@demo-shop/api';
import { CartFacade } from '@demo-shop/cart';
import { UserNavigationComponent } from './user-navigation.component';
import { mockUser, UserFacade } from '@demo-shop/user';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { AuthFacade } from '@demo-shop/auth';

const mockUserFacade = (user?: UserResponse) => ({
  getCurrentUser: () => signal(user),
});

const mockAuthFacade = {
  login: () => {
    return;
  },
  logout: () => {
    return;
  },
  register: () => {
    return;
  },
};

const mockCartFacade = {
  loadShoppingSession: () => {
    return;
  },
};

const meta: Meta<UserNavigationComponent> = {
  component: UserNavigationComponent,
  title: 'Navigation/UserNavigationComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule),
        { provide: AuthFacade, useValue: mockAuthFacade },
        { provide: CartFacade, useValue: mockCartFacade },
      ],
    }),
  ],
  argTypes: {
    isSmallScreen: {
      control: {
        type: 'boolean',
      },
    },
    login: { action: 'login', control: false },
    register: { action: 'register', control: false },
    logout: { action: 'logout', control: false },
  },
  parameters: {
    controls: {
      exclude: ['#authFacade', '#userFacade', 'currentUser', '#cartFacade'],
    },
  },
};
export default meta;
type Story = StoryObj<UserNavigationComponent>;

export const Unauthenticated: Story = {
  args: {
    isSmallScreen: false,
  },
  decorators: [
    applicationConfig({
      providers: [{ provide: UserFacade, useValue: mockUserFacade() }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const signIn = within(canvasElement).getByText('Sign in');
    const register = within(canvasElement).getByText('Register');
    expect(signIn).toBeInTheDocument();
    expect(register).toBeInTheDocument();
  },
};

export const Authenticated: Story = {
  args: {
    isSmallScreen: false,
  },
  decorators: [
    applicationConfig({
      providers: [{ provide: UserFacade, useValue: mockUserFacade(mockUser) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const signOut = within(canvasElement).getByText('Sign out');
    expect(signOut).not.toBeDisabled();
  },
};
