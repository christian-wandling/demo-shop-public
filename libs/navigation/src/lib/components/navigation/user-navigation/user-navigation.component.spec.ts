import { ComponentFixture, TestBed } from '@angular/core/testing';
import { UserNavigationComponent } from './user-navigation.component';
import { AuthFacade } from '@demo-shop/auth';
import { UserFacade } from '@demo-shop/user';
import { signal } from '@angular/core';
import { UserDTO } from '@demo-shop/api';
import { CartFacade } from '@demo-shop/cart';

describe('UserNavigationComponent', () => {
  const user = signal<UserDTO>({
    address: {
      street: 'street',
      apartment: 'apartment',
      city: 'city',
      region: 'region',
      zip: 'zip',
      country: 'country',
    },
    email: 'email',
    firstname: 'firstname',
    id: 1,
    lastname: 'lastname',
    phone: 'phone',
  });

  let component: UserNavigationComponent;
  let fixture: ComponentFixture<UserNavigationComponent>;
  let authFacade: AuthFacade;
  let cartFacade: CartFacade;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [UserNavigationComponent],
      providers: [
        { provide: AuthFacade, useValue: { login: jest.fn(), logout: jest.fn(), register: jest.fn() } },
        { provide: UserFacade, useValue: { getCurrentUser: jest.fn().mockReturnValue(user) } },
        { provide: CartFacade, useValue: { loadShoppingSession: jest.fn() } },
      ],
    }).compileComponents();

    authFacade = TestBed.inject(AuthFacade);
    cartFacade = TestBed.inject(CartFacade);
    fixture = TestBed.createComponent(UserNavigationComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should handle isSmallScreen input', () => {
    // Test default value
    expect(component.isSmallScreen()).toBe(false);

    // Test setting to true
    fixture.componentRef.setInput('isSmallScreen', true);
    expect(component.isSmallScreen()).toBe(true);
  });

  it('should call auth facade login and load shopping session', async () => {
    await component.login();

    expect(authFacade.login).toHaveBeenCalled();
    expect(cartFacade.loadShoppingSession).toHaveBeenCalled();
  });

  it('should call auth facade register and load shopping session', async () => {
    await component.register();

    expect(authFacade.register).toHaveBeenCalled();
    expect(cartFacade.loadShoppingSession).toHaveBeenCalled();
  });

  it('should call auth facade logout', async () => {
    await component.logout();

    expect(authFacade.logout).toHaveBeenCalled();
  });
});
