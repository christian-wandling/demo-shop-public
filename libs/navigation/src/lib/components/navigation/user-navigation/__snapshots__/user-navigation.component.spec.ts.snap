// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserNavigationComponent snapshots should match snapshot for small screens 1`] = `
<div
  id="root7"
>
  <div
    class="flow-root"
  >
    <a
      class="-m-2 block p-2 font-medium text-xl text-gray-400"
    >
       firstname lastname 
    </a>
  </div>
  <div
    class="flow-root"
  >
    <a
      class="-m-2 block p-2 font-medium text-xl text-gray-700 hover:text-indigo-500 cursor-pointer"
    >
       Sign out 
    </a>
  </div>
  
  
  
  
  
</div>
`;

exports[`UserNavigationComponent snapshots should match snapshot with default state 1`] = `
<div
  id="root6"
>
  <a
    class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"
  >
     firstname lastname 
  </a>
  <span
    aria-hidden="true"
    class="h-6 w-px bg-gray-200"
  />
  <a
    class="text-sm font-medium text-gray-700 hover:text-gray-800 cursor-pointer"
  >
     Sign out 
  </a>
  
  
  
  
</div>
`;
